File: mock_backend/app/schemas/interview.py
import uuid
from pydantic import BaseModel, Field
from typing import Optional, List, Any, Literal
from datetime import datetime


# ─── Request Schemas ─────────────────────────────────────────────────────────

class InterviewSessionQuestionCreate(BaseModel):
    """Schema for creating a session question with optional overrides."""
    question_id: Optional[uuid.UUID] = None
    custom_text: Optional[str] = None
    order: int

class ScheduleInterviewRequest(BaseModel):
    """Request body for POST /admin/interviews/schedule"""
    candidate_id: str = Field(..., description="UUID of the candidate user")
    template_id: str = Field(..., description="UUID of the interview template")
    scheduled_at: datetime = Field(..., description="Future UTC datetime for the interview")
    questions: Optional[List[InterviewSessionQuestionCreate]] = Field(None, description="Optional custom questions to override template defaults")
    draft_interview_id: Optional[str] = Field(None, description="UUID of the draft interview to finalize")


class RescheduleInterviewRequest(BaseModel):
    """Request body for PUT /admin/interviews/{id}/reschedule"""
    scheduled_at: datetime = Field(..., description="New future UTC datetime for the interview")


class CancelInterviewRequest(BaseModel):
    """Request body for PUT /admin/interviews/{id}/cancel"""
    reason: Optional[str] = Field(None, description="Optional cancellation reason for audit")


# ─── Response Schemas ─────────────────────────────────────────────────────────

class CuratedQuestionItem(BaseModel):
    """
    A single question in the curated payload.

    Common fields are required for all types.
    Type-specific config blocks are Optional — only the relevant one
    will be populated depending on question_type:
      - "static"         → evaluation_mode + source
      - "conversational" → conversation_config
      - "coding"         → coding_config
    """
    question_id: str
    question_type: Literal["static", "conversational", "coding"]
    order: int
    prompt: str
    text: Optional[str] = None
    difficulty: Literal["easy", "medium", "hard"]
    time_limit_sec: int

    # static-only fields
    evaluation_mode: Optional[str] = None   # e.g. "text"
    source: Optional[str] = None            # e.g. "question_bank"

    # conversational-only fields
    conversation_config: Optional[dict] = None
    # Expected shape:
    # { "follow_up_depth": int, "ai_model": str, "evaluation_mode": str }

    # coding-only fields
    coding_config: Optional[dict] = None
    # Expected shape:
    # { "language": str, "starter_code": str,
    #   "test_cases": [{"input": str, "expected_output": str}],
    #   "execution_timeout_sec": int }

    class Config:
        extra = "allow"   # forward-compatible: extra keys from AI agent won't break deserialization


class TechnicalSection(BaseModel):
    questions: List[CuratedQuestionItem]

class CodingProblemItem(BaseModel):
    problem_id: str
    title: str
    difficulty: str
    description: Optional[str] = None
    starter_code: Optional[dict] = None

class CodingSection(BaseModel):
    problems: List[CodingProblemItem]

class ConversationalSection(BaseModel):
    rounds: int

class CuratedQuestionsPayload(BaseModel):
    template_id: str
    generated_from: dict
    generated_at: datetime
    generation_method: str
    questions: Optional[List[CuratedQuestionItem]] = None
    technical_section: Optional[TechnicalSection] = None
    coding_section: Optional[CodingSection] = None
    conversational_section: Optional[ConversationalSection] = None


class InterviewTemplateResponse(BaseModel):
    id: uuid.UUID
    title: str
    role_name: Optional[str] = None
    description: Optional[str] = None
    is_active: bool
    is_default_for_role: bool
    settings: Optional[dict] = None
    technical_config: Optional[dict] = None
    coding_config: Optional[dict] = None
    conversational_config: Optional[dict] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class InterviewTemplateCreate(BaseModel):
    title: str
    role_name: Optional[str] = None
    description: Optional[str] = None
    is_active: bool = True
    is_default_for_role: bool = False
    settings: Optional[dict] = None
    technical_config: Optional[dict] = None
    coding_config: Optional[dict] = None
    conversational_config: Optional[dict] = None


class InterviewTemplateUpdate(BaseModel):
    title: Optional[str] = None
    role_name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    is_default_for_role: Optional[bool] = None
    settings: Optional[dict] = None
    technical_config: Optional[dict] = None
    coding_config: Optional[dict] = None
    conversational_config: Optional[dict] = None


class ScheduleInterviewResponse(BaseModel):
    """Response body for POST /admin/interviews/schedule"""
    id: str
    candidate_id: str
    template_id: str
    assigned_by: str
    status: str
    scheduled_at: datetime
    curated_questions: CuratedQuestionsPayload
    created_at: datetime

    class Config:
        from_attributes = True


class CancelInterviewResponse(BaseModel):
    """Response body for PUT /admin/interviews/{id}/cancel"""
    id: str
    status: str
    cancelled_at: datetime
    reason: Optional[str] = None

    class Config:
        from_attributes = True

class RescheduleInterviewResponse(BaseModel):
    """Response body for PUT /admin/interviews/{id}/reschedule"""
    id: str
    status: str
    scheduled_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/interview.ts
// ─── Interview Scheduling API Types ──────────────────────────────────────────

export type InterviewStatus =
    | 'scheduled'
    | 'in_progress'
    | 'completed'
    | 'cancelled'
    | 'draft';

export interface InterviewTemplate {
    id: string;
    title: string;
    role_name?: string | null;
    description?: string | null;
    is_active: boolean;
    is_default_for_role: boolean;
    created_at: string;
    updated_at: string;
    settings?: Record<string, any> | null;
    technical_config?: Record<string, any> | null;
    coding_config?: Record<string, any> | null;
    conversational_config?: Record<string, any> | null;
}

export interface InterviewTemplateCreate {
    title: string;
    role_name?: string;
    description?: string;
    is_active?: boolean;
    is_default_for_role?: boolean;
    settings?: Record<string, any>;
    technical_config?: Record<string, any> | null;
    coding_config?: Record<string, any> | null;
    conversational_config?: Record<string, any> | null;
}

export interface InterviewTemplateUpdate {
    title?: string;
    role_name?: string;
    description?: string;
    is_active?: boolean;
    is_default_for_role?: boolean;
    settings?: Record<string, any>;
    technical_config?: Record<string, any> | null;
    coding_config?: Record<string, any> | null;
    conversational_config?: Record<string, any> | null;
}

export interface CandidateInterview {
    id: string;
    candidate_id: string;
    template_id: string;
    assigned_by: string;
    status: InterviewStatus;
    scheduled_at: string;
    created_at: string;
}

// Extend the existing CandidateResponse with interview info
export interface CandidateWithInterview {
    id: string;
    username: string;
    email: string;
    is_active: boolean;
    login_disabled: boolean;
    created_at: string;
    interview?: CandidateInterview | null;  // null = no interview yet
}

// ─── Request / Response shapes ───────────────────────────────────────────────

export interface InterviewSessionQuestionCreate {
    question_id?: string;
    custom_text?: string;
    order: number;
}

export interface ScheduleInterviewRequest {
    candidate_id: string;
    template_id: string;
    scheduled_at: string; // ISO 8601 UTC
    questions?: InterviewSessionQuestionCreate[];
    draft_interview_id?: string;
}

export interface TemplatePreviewQuestion {
    question_id: string;
    text: string;
    originalText?: string;
    difficulty: string;
    category: string;
}

export interface TemplatePreviewResponse {
    interview_id?: string;
    technical_section: { questions: TemplatePreviewQuestion[] };
    coding_section: { problems: { problem_id: string; title: string; difficulty: string }[] };
    conversational_section: { rounds: number; description: string };
}

export interface RescheduleInterviewRequest {
    scheduled_at: string; // ISO 8601 UTC
}

export interface CancelInterviewRequest {
    reason?: string;
}

export interface ScheduleInterviewResponse {
    id: string;
    candidate_id: string;
    template_id: string;
    assigned_by: string;
    status: 'scheduled';
    scheduled_at: string;
    created_at: string;
}

export interface RescheduleInterviewResponse {
    id: string;
    status: 'scheduled';
    scheduled_at: string;
    updated_at: string;
}

export interface CancelInterviewResponse {
    id: string;
    status: 'cancelled';
    cancelled_at: string;
    reason?: string;
}

// ─── Structured API error ────────────────────────────────────────────────────

export interface SchedulingApiError {
    status: number;
    detail: string;
}
